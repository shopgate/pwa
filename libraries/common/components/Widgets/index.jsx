import React, { Component } from 'react';
import PropTypes from 'prop-types';
import isEqual from 'lodash/isEqual';
import WidgetGrid from './components/WidgetGrid';
import shouldShowWidget from './helpers/shouldShowWidget';

const WIDGET_GRID_TYPE = '@shopgate/commerce-widgets/widget-grid';
const GRID_COLUMNS = 12; // One grid row has 12 columns. // TODO: is it deprecated since css grid?

/**
 * Creates a grid wrapper for widget(s).
 * @param {string} key The unique key.
 * @param {Array} config Array of widgets to be placed in the grid.
 * @param {Array} components The component definitions for the widgets.
 * @returns {JSX} The wrapper.
 */
const createGridWrapper = (key, config, components) => (
  React.createElement(
    WidgetGrid,
    {
      key,
      config,
      components,
    }
  )
);

/**
 * The widgets component.
 */
class Widgets extends Component {
  static propTypes = {
    components: PropTypes.shape(),
    widgets: PropTypes.arrayOf(PropTypes.shape()),
  };

  static defaultProps = {
    components: null,
    widgets: null,
  };

  /**
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);
    this.autoReloadInterval = undefined;
    if (this.hasSchedulableWidgets()) {
      this.startAutoRerender();
    }
  }

  /**
   * @param {Object} nextProps The next component props.
   * @return {JSX}
   */
  shouldComponentUpdate(nextProps) {
    if (!isEqual(this.props.components, nextProps.components)) {
      return true;
    }

    if (!isEqual(this.props.widgets, nextProps.widgets)) {
      return true;
    }

    return false;
  }

  /**
   * Component will unmount lifecycle method.
   */
  componentWillUnmount() {
    this.stopAutoRerender();
  }

  /**
   * Checks if any widget is schedulable.
   * @param {Array} widgets Array of widgets.
   * @returns {boolean}
   */
  hasSchedulableWidgets() {
    return (this.props.widgets || []).some(widget => widget.settings.plan);
  }

  /**
   * Sets auto re-render.
   */
  startAutoRerender() {
    const minutesToRoundedHour = 60 - new Date().getMinutes();
    const nextRerenderIn = minutesToRoundedHour * 60000;
    this.autoReloadTimeout = setTimeout(() => this.doAutoRerender(), nextRerenderIn);
  }

  /**
   * Stops auto re-render. Must be called before component is unmounted to avoid
   * memory leak.
   */
  stopAutoRerender() {
    clearTimeout(this.autoReloadTimeout);
  }

  /**
   * Forces re-render and sets another timeout for next cycle.
   */
  doAutoRerender() {
    this.forceUpdate();
    this.startAutoRerender();
  }

  /**
   * Create array of elements from widget configuration.
   * @returns {Array} Array of JSX elements.
   */
  createArrayOfElements() {
    const { widgets = [], components } = this.props;

    return widgets
      .filter(widget => shouldShowWidget(widget.settings))
      .map((widget, index) => {
        if (!components[widget.type] && widget.type !== WIDGET_GRID_TYPE) {
          return null;
        }

        const key = `w${index}`;

        if (widget.type === WIDGET_GRID_TYPE) {
          // If it's a grid just create it and pass the child widgets.
          return createGridWrapper(key, widget.settings.widgets, components);
        }

        if (widget.height) {
          // If it has a definite height wrap the widget in a grid.
          return createGridWrapper(
            key,
            [{
              ...widget,
              col: 0,
              row: 0,
              width: GRID_COLUMNS,
            }],
            components
          );
        }

        // In all other cases just create and return the widget component.
        return React.createElement(
          components[widget.type],
          {
            ...widget,
            key,
          }
        );
      });
  }

  /**
   * @return {JSX}
   */
  render() {
    const { widgets, components } = this.props;

    if (!widgets) {
      return null;
    }

    return (
      <div className="common__widgets">{this.createArrayOfElements(widgets, components)}</div>
    );
  }
}

export default Widgets;
