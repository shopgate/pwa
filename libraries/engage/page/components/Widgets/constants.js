/**
 * List of allowed origins for cms page preview iFrame communication.
 */
export const ALLOWED_PAGE_PREVIEW_ORIGINS = [
  'next.admin.shopgatedev.com',
  'next.admin.shopgatepg.com',
  'next.admin.shopgate.com',
  'next.us.admin.shopgate.com',
  'http://localhost:1337',
];

// Whether to consider vertical margins when calculating the overlay position.
export const CONSIDER_CONTAINER_MARGINS_DEFAULT = false;
