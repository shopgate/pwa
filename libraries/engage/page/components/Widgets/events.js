import { useEffect } from 'react';

/**
 * @typedef {"highlight-widget"|"widget-clicked"} WidgetPreviewEventName
 */

/**
 * @typedef {Object} WidgetPreviewEventDetail
 * @property {string} widgetCode The code of the widget related to the event.
 * @property {any} [payload] Optional payload data related to the event.
 */

/**
 * @callback WidgetPreviewEventHandler
 * @param {CustomEvent<WidgetPreviewEventDetail>} event The custom event dispatched for the widget
 * preview.
 */

/**
 * Hook to listen for widget preview events.
 * These events are dispatched in the context of iFrame communication at the widget preview.
 * @param {WidgetPreviewEventName} eventName Name of the listened event
 * @param {WidgetPreviewEventHandler} handler A callback function to handle the event
 */
export const useWidgetPreviewEvent = (eventName, handler) => {
  useEffect(() => {
    window.addEventListener(`widget-preview-${eventName}`, handler);

    return () => {
      window.removeEventListener(`widget-preview-${eventName}`, handler);
    };
  }, [eventName, handler]);
};

/**
 * Dispatches widget preview related events.
 * Used to trigger iFrame postMessage events to the parent window or to react on incoming
 * postMessage events from the parent window.
 * @param {WidgetPreviewEventName} eventName Name of the event to dispatch
 * @param {string} widgetCode Code of the widget to dispatch the event for
 * @param {Object} [payload] Optional payload to include with the event
 */
export const dispatchWidgetPreviewEvent = (eventName, widgetCode, payload = null) => {
  const event = new CustomEvent(`widget-preview-${eventName}`, {
    detail: {
      widgetCode,
      payload,
    },
  });

  window.dispatchEvent(event);
};
