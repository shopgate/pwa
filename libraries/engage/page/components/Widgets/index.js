export { default } from './Widgets';
export { WidgetContext } from './WidgetContext';
