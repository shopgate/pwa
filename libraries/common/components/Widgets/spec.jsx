import React from 'react';
import { mount } from 'enzyme';
import Widgets from './index';

jest.useFakeTimers();

jest.mock('react', () => ({
  ...require.requireActual('react'),
  Suspense: function Suspense({ children }) { return children; },
}));

/**
 * A mock Image component.
 * @returns {JSX}
 */
const Image = () => <img alt="" />;
/* eslint-disable react/prop-types */
/**
 * A mock WidgetGrid component.
 * @param {Array} children Array of children.
 * @returns {JSX}
 */
const WidgetGrid = ({ children }) => <div className="widget-grid">{children}</div>;
/* eslint-enable react/prop-types */

const components = {
  '@shopgate/commerce-widgets/image': Image,
  '@shopgate/commerce-widgets/widget-grid': WidgetGrid,
};

describe('<Widgets />', () => {
  it('should render a grid if height is defined', () => {
    const widgets = [{
      col: 0,
      row: 0,
      width: 12,
      height: 3,
      settings: {
        id: 83535,
        image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
      },
      type: '@shopgate/commerce-widgets/image',
    }];

    const wrapper = mount((
      <Widgets
        components={components}
        widgets={widgets}
      />
    ));

    expect(wrapper.find('WidgetGrid').exists()).toBe(true);
  });

  it('should not wrap a widget which is not a grid and has no height', () => {
    const widgets = [{
      col: 0,
      row: 0,
      width: 12,
      settings: {
        id: 83535,
        image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
      },
      type: '@shopgate/commerce-widgets/image',
    }];

    const wrapper = mount((
      <Widgets
        components={components}
        widgets={widgets}
      />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('WidgetGrid').exists()).toBe(false);
  });

  it('should render a grid if the widget is of type grid', () => {
    const widgets = [{
      type: '@shopgate/commerce-widgets/widget-grid',
      settings: {
        widgets: [
          {
            col: 0,
            row: 0,
            width: 12,
            height: 5,
            settings: {
              id: '84961',
              alt: '',
              image: 'https://data.shopgate.com/shop_widget_images/23836/aedc545959f55e3f73851eca0ed40a75.min.jpeg',
              link: '/category/',
            },
            type: '@shopgate/commerce-widgets/image',
          },
        ],
      },
    }];

    const wrapper = mount((
      <Widgets
        components={components}
        widgets={widgets}
      />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('WidgetGrid').exists()).toBe(true);
  });

  it('should render only one widget when the second one is not published and third one is invalid', () => {
    const widgets = [
      {
        col: 0,
        row: 0,
        width: 12,
        settings: {
          id: 835351,
          image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
        },
        type: '@shopgate/commerce-widgets/image',
      },
      {
        col: 0,
        row: 0,
        width: 12,
        settings: {
          id: 835352,
          image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
        },
        type: '@shopgate/commerce-widgets/imagefoo',
      },
      {
        col: 0,
        row: 0,
        width: 12,
        settings: {
          published: false,
          id: 835353,
          image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
        },
        type: '@shopgate/commerce-widgets/image',
      },
    ];

    const wrapper = mount((
      <Widgets
        components={components}
        widgets={widgets}
      />
    ));

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('img').length).toBe(1);
  });

  it('should schedule a re-render when widget is scheduled', () => {
    const minutesToNextFullHour = 60 - new Date().getMinutes();
    const msToNextFullHour = minutesToNextFullHour * 60000;
    const scheduledFromMs = (Date.now() + msToNextFullHour) - 1;
    const scheduledToMs = Date.now() + minutesToNextFullHour + 1000;
    /* eslint-disable camelcase */
    const widgets = [
      {
        col: 0,
        row: 0,
        width: 12,
        settings: {
          id: 835351,
          image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
          published: true,
          plan: true,
          planDate: {
            valid_from: new Date(scheduledFromMs).toISOString(),
            valid_to: new Date(scheduledToMs).toISOString(),
          },
        },
        type: '@shopgate/commerce-widgets/image',
      },
    ];
    /* eslint-enable camelcase */
    const wrapper = mount((
      <Widgets
        components={components}
        widgets={widgets}
      />
    ));
    wrapper.instance().forceUpdate = jest.fn();
    expect(wrapper.find(Image).exists()).toBe(false);
    jest.advanceTimersByTime(msToNextFullHour);
    expect(wrapper.instance().forceUpdate).toHaveBeenCalledTimes(1);
    // In real life next timeout should be in 60 minutes.
    // This test has same Date and fake timers.
    jest.advanceTimersByTime(msToNextFullHour);
    expect(wrapper.instance().forceUpdate).toHaveBeenCalledTimes(2);
    wrapper.instance().componentWillUnmount();
    expect(clearTimeout).toHaveBeenCalled();
  });

  it('should render only wrapper when widgets array is empty', () => {
    const widgets = [];
    const wrapper = mount((
      <Widgets
        components={components}
        widgets={widgets}
      />
    ));
    expect(wrapper.find('Image').exists()).toBe(false);
  });

  it('should render null when no widgets are passed', () => {
    const wrapper = mount((
      <Widgets
        components={components}
      />
    ));
    expect(wrapper.html()).toBe(null);
  });

  it('should check settings of child widgets inside widget-grid', () => {
    const widgets = [
      {
        height: 2,
        id: 'index-5-@shopgate/commerce-widgets/widget-grid',
        type: '@shopgate/commerce-widgets/widget-grid',
        settings: {
          widgets: [
            {
              col: 0,
              row: 0,
              height: 2,
              width: 2,
              settings: {
                id: 835351,
                image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
                published: true,
                plan: false,
              },
              type: '@shopgate/commerce-widgets/image',
            },
            {
              col: 2,
              row: 0,
              height: 2,
              width: 2,
              settings: {
                id: 835352,
                image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
                published: false,
                plan: false,
              },
              type: '@shopgate/commerce-widgets/image',
            },
          ],
        },
      },
    ];
    const wrapper = mount((
      <Widgets
        components={components}
        widgets={widgets}
      />
    ));
    expect(wrapper.find('img').length).toBe(1);
  });
});
