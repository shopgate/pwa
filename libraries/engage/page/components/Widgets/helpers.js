import { PAGE_PREVIEW_PATTERN } from '@shopgate/engage/page/constants';

/**
 * Retrieves the scroll container for the current page. Depending on the PWA mode this can be
 * a scrollable article element or the window.
 * @returns {HTMLElement|null}
 */
export const getScrollContainer = () => document.querySelector(`.route__${PAGE_PREVIEW_PATTERN.replace(/^\/+/, '')}`);

/**
 * @typedef {Object} ScheduledParams
 * @param {string} [from] The start date of the scheduling in ISO format.
 * @param {string} [to] The end date of the scheduling in ISO format.
 * @param {number} [timezoneOffset] The timezone offset in minutes. If not provided, the local
 * timezone offset will be used.
 */

/**
 * @typedef {Object} ScheduledStatus
 * @param {boolean} isScheduled Indicates if the widget is scheduled.
 * @param {boolean} isHidden Indicates if the widget is currently hidden based on the scheduling.
 * @param {boolean} isExpired Indicates if the scheduled time frame has expired.
 */

/**
 * Retrieves the scheduling status of a widget based on the provided parameters.
 * @param {ScheduledParams} params The parameters for the function.
 * @returns {ScheduledStatus} An object containing the scheduling status.
 */
export function checkScheduled({ from, to, timezoneOffset } = {}) {
  const now = new Date();

  // Convert current time to provided or local timezone
  const localOffset = timezoneOffset ?? -now.getTimezoneOffset(); // in minutes
  const offsetMs = localOffset * 60 * 1000;
  const localNow = new Date(now.getTime() + offsetMs);

  const fromDate = from ? new Date(from) : null;
  const toDate = to ? new Date(to) : null;

  const inTimeframe = (!fromDate || localNow >= new Date(fromDate.getTime() + offsetMs)) &&
                      (!toDate || localNow <= new Date(toDate.getTime() + offsetMs));

  const isExpired = !!toDate && localNow > new Date(toDate.getTime() + offsetMs);
  const isScheduled = !!fromDate || !!toDate;
  return {
    isScheduled,
    isHidden: !inTimeframe,
    isExpired,
  };
}
