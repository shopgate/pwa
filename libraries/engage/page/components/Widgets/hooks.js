import {
  useEffect, useCallback, useRef, useContext, useMemo,
} from 'react';
import { logger } from '@shopgate/engage/core/helpers';
import { useDispatch } from 'react-redux';
import { useRoute } from '@shopgate/engage/core/hooks';
import { receivePageConfigV2 } from '@shopgate/engage/page/action-creators';
import { PAGE_PREVIEW_SLUG } from '@shopgate/engage/page/constants';
import {
  ALLOWED_PAGE_PREVIEW_ORIGINS,
  CONSIDER_CONTAINER_MARGINS_ON_SCROLL_DEFAULT,
} from './constants';
import { getScrollContainer } from './helpers';
import { WidgetsPreviewContext } from './WidgetsPreviewContext';
import {
  dispatchWidgetPreviewEvent,
  useWidgetPreviewEvent,
} from './events';

/**
 * @typedef {Object} MessageData
 * @property {string} type Identifier for the kind of message
 * @property {any} [payload] Optional data payload for this message
 */

/**
 * @typedef {Object} IframeMessengerResult
 * @property {function(MessageData, string=): void} sendToParent
 *   - Send data up to window.parent. If targetOrigin is omitted, uses the
 *     most recently seen origin (from an incoming message). If none seen yet,
 *     falls back to parentOrigins[0] or "*".
 */

/**
 * Hook for postMessage communication when your component is inside an iframe.
 *
 * Listens on window for "message" events from any origin in parentOrigins,
 * and only calls onMessage(data, rawEvent) if both origin and source match.
 *
 * @param {function(MessageData, any): void} onMessage
 *   Callback invoked when a trusted message arrives. Receives data and the
 *   raw event (so you can inspect origin, source, etc.).
 * @param {string[]} parentOrigins
 *   Array of allowed parent origin strings (e.g.
 *   ['https://a.example.com','https://b.example.com']).
 * @returns {IframeMessengerResult}
 *   An object with a single method:
 *   • sendToParent(data, [targetOrigin]): void
 *     – Posts data up to window.parent. By default it uses the most recently
 *       seen origin (from an incoming message). If none, uses parentOrigins[0].
 */
function useIframeMessenger(onMessage, parentOrigins) {
  // Keep a ref to the latest onMessage callback so the listener always has it.
  const onMessageRef = useRef(onMessage);
  useEffect(() => {
    onMessageRef.current = onMessage;
  }, [onMessage]);

  // Keep track of the last allowed origin we heard from
  const lastOriginRef = useRef(null);

  // Attach / detach the "message" listener.
  useEffect(() => {
    /**
     * Handler for incoming postMessage events.
     * @param {any} rawEvent  – The original MessageEvent object.
     */
    function handler(rawEvent) {
      if (rawEvent?.data?.type === 'webpackOk') {
        onMessageRef.current({
          type: 'requestPageConfig',
        }, rawEvent);

        return;
      }

      // Only proceed if the origin is in our whitelist.
      if (!parentOrigins.includes(rawEvent.origin)) return;
      // Ensure the message actually came from window.parent.
      if (rawEvent.source !== window.parent) return;

      // Record this origin as most recently seen.
      lastOriginRef.current = rawEvent.origin;

      // Forward the event.data and the raw event to the callback.
      onMessageRef.current(rawEvent.data, rawEvent);
    }

    window.addEventListener('message', handler);
    return () => {
      window.removeEventListener('message', handler);
    };
  }, [parentOrigins]);

  /**
   * Send a message up to the parent window.
   * @param {MessageData} data       - The data object to post.
   * @param {string}      [targetOrigin]
   *   Optional override for the origin to post to. Must be one of
   *   parentOrigins. If omitted, uses the last seen origin (lastOriginRef),
   *   or parentOrigins[0], or "*" if array is empty.
   */
  const sendToParent = useCallback(
    (data, targetOrigin) => {
      // Determine which origin to use: explicit, then last seen, then first, then "*".
      const originToUse =
        typeof targetOrigin === 'string'
          ? targetOrigin
          : lastOriginRef.current || new URL(document.referrer).origin || parentOrigins[0] || '*';

      if (!originToUse) {
        logger.warn(
          'useIframeMessenger: no targetOrigin available. ' +
            'Provide parentOrigins or pass targetOrigin.'
        );
        return;
      }

      window.parent.postMessage(data, originToUse);
    },
    [parentOrigins]
  );

  return { sendToParent };
}

/**
 * Hook to handle communication with the parent window in a page preview iframe.
 * @param {boolean} isActive Whether the preview communication is active.
 */
export const usePreviewIframeCommunication = (isActive = false) => {
  const dispatch = useDispatch();

  const { query: { considerContainerMarginsOnScroll } } = useRoute();

  // Detect if container margins should be considered at scroll to widget.
  const considerVerticalMargins = useMemo(() => {
    if (!considerContainerMarginsOnScroll) {
      return CONSIDER_CONTAINER_MARGINS_ON_SCROLL_DEFAULT;
    }

    return considerContainerMarginsOnScroll === 'true';
  }, [considerContainerMarginsOnScroll]);

  const { sendToParent } = useIframeMessenger((data) => {
    if (data.type === 'receivePageConfig') {
      // Page preview config received from the parent window.
      dispatch(receivePageConfigV2({
        type: 'cms',
        slug: PAGE_PREVIEW_SLUG,
        data: data.payload,
      }));
    } else if (data.type === 'scrollToWidget' && data.payload?.widgetCode) {
      // Parent window requested to scroll to a specific widget.
      const scrollContainer = getScrollContainer();
      const target = document.getElementById(data.payload.widgetCode);

      if (scrollContainer && target) {
        let marginTop = 0;

        if (considerVerticalMargins) {
          const styles = window.getComputedStyle(target);
          marginTop = parseFloat(styles.marginTop);
        }

        const containerTop = scrollContainer.getBoundingClientRect().top;
        const targetTop = target.getBoundingClientRect().top;
        const scrollOffset = targetTop - containerTop + scrollContainer.scrollTop - marginTop;
        const maxScrollTop = scrollContainer.scrollHeight - scrollContainer.clientHeight;
        const actualScrollTop = Math.min(scrollOffset, maxScrollTop);

        // Register the target element as the active widget.
        dispatchWidgetPreviewEvent('set-active-widget-id', data.payload.widgetCode);

        /**
         * Callback to highlight the widget after scrolling.
         */
        const highlightWidget = () => {
          dispatchWidgetPreviewEvent('highlight-widget', data.payload.widgetCode);
        };

        // Add listener to onScrollEnd if available, otherwise use scroll event.
        if ('onscrollend' in scrollContainer) {
          /**
           * Callback for the scrollend event.
           */
          const onEnded = () => {
            scrollContainer.removeEventListener('scrollend', onEnded);
            highlightWidget();
          };
          scrollContainer.addEventListener('scrollend', onEnded, { once: true });
          scrollContainer.scrollTo({
            top: actualScrollTop,
            behavior: 'smooth',
          });
          return;
        }

        // Fallback: listen for scroll events until scrollTop ≈ actualScrollTop

        /**
         * Callback for the scroll event.
         */
        const onScroll = () => {
          // Allow a 1 px leeway for subpixel rendering
          if (Math.abs(scrollContainer.scrollTop - actualScrollTop) < 1) {
            scrollContainer.removeEventListener('scroll', onScroll);
            highlightWidget();
          }
        };

        scrollContainer.addEventListener('scroll', onScroll);
        scrollContainer.scrollTo({
          top: actualScrollTop,
          behavior: 'smooth',
        });
      }
    }
  }, ALLOWED_PAGE_PREVIEW_ORIGINS);

  useWidgetPreviewEvent('widget-clicked', (e) => {
    if (!isActive) {
      return;
    }

    sendToParent({
      type: 'widgetClicked',
      payload: {
        widgetCode: e.detail.widgetCode,
      },
    });
  });
};

/**
 * @typedef {import('./WidgetsPreviewContext.js').WidgetsPreviewContextType}
 * WidgetsPreviewContextType
 */

/**
 * The useWidgetsPreview hook provides access to the context that is wrapped around the Widgets
 * component when it's rendered in preview mode.
 * @returns {WidgetsPreviewContextType} The widget context.
 */
export const useWidgetsPreview = () => useContext(WidgetsPreviewContext);
