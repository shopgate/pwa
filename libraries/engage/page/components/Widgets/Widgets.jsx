import React, { useMemo, useRef } from 'react';
import PropTypes from 'prop-types';
import { useRoute, useThemeWidgets } from '@shopgate/engage/core/hooks';
import { PAGE_PREVIEW_PATTERN } from '@shopgate/engage/page/constants';
import { ConditionalWrapper } from '@shopgate/engage/components';
import WidgetsPreviewProvider from './WidgetsPreviewProvider';
import Widget from './Widget';
import Overlay from './Overlay';
import { usePreviewIframeCommunication } from './hooks';

/**
 * @typedef {Object} WidgetDefinitionVisibility
 * @property {boolean} isHidden Whether the widget is hidden.
 * @property {string} scheduleStartDate Start date for scheduled widgets.
 * @property {string} scheduleEndDate End date for scheduled widgets.
 */

/**
 * @typedef {Object} WidgetDefinitionLayout
 * @property {number} marginTop Top margin for the widget.
 * @property {number} marginBottom Bottom margin for the widget.
 * @property {number} marginLeft Left margin for the widget.
 * @property {number} marginRight Right margin for the widget.
 */

/**
 * @typedef {Object} WidgetDefinition
 * @property {string} code Unique code for the widget.
 * @property {string} widgetConfigDefinitionCode Name of the widget.
 * @property {Object} widgetConfig Individual configuration for the widget.
 * @property {WidgetDefinitionVisibility} visibility Visibility settings for the widget.
 * @property {WidgetDefinitionLayout} layout Layout settings for the widget.
 */

const PLACEHOLDER_COMPONENT = '@shopgate/widgetsInternal/Placeholder';

/**
 * The Widgets component renders a list of widgets.
 * @param {Object} props The component props.
 * @param {Array<WidgetDefinition>} props.widgets The list of widgets to render.
 * @returns {JSX.Element}
 */
const Widgets = ({
  widgets: widgetsProp = [],
}) => {
  const { pattern } = useRoute();
  const widgetsRef = useRef(null);
  const isPreview = pattern === PAGE_PREVIEW_PATTERN;
  const widgetComponents = useThemeWidgets('v2');

  usePreviewIframeCommunication(isPreview);

  // Create sanitized widgets array that only includes widgets with valid components.
  const widgets = useMemo(() => {
    if (isPreview) {
      // All widgets are allowed in preview mode.
      return widgetsProp;
    }

    // Remove widgets that do not have a valid component.
    return widgetsProp.filter(widget => !!widgetComponents[widget.widgetConfigDefinitionCode]);
  }, [isPreview, widgetComponents, widgetsProp]);

  if (!Array.isArray(widgets) || widgets.length === 0) {
    return null;
  }
  return (
    <ConditionalWrapper
      condition={isPreview}
      wrapper={children => (
        <WidgetsPreviewProvider>
          {children}
          <Overlay containerRef={widgetsRef} />
        </WidgetsPreviewProvider>
      )}
    >
      <div className="engage__widgets" ref={widgetsRef}>
        {widgets.map((widget) => {
          const component = widgetComponents[widget.widgetConfigDefinitionCode] ||
          widgetComponents[PLACEHOLDER_COMPONENT];

          return <Widget
            key={widget.code}
            definition={widget}
            isPreview={isPreview}
            component={component}
          />;
        })}
      </div>
    </ConditionalWrapper>
  );
};

Widgets.propTypes = {
  widgets: PropTypes.arrayOf(PropTypes.shape()),
};

Widgets.defaultProps = {
  widgets: null,
};

export default Widgets;
