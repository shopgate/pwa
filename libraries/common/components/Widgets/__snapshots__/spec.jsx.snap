// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Widgets /> should not wrap a widget which is not a grid and has no height 1`] = `
<Widgets
  components={
    Object {
      "@shopgate/commerce-widgets/image": [Function],
      "@shopgate/commerce-widgets/widget-grid": [Function],
    }
  }
  widgets={
    Array [
      Object {
        "col": 0,
        "row": 0,
        "settings": Object {
          "id": 83535,
          "image": "https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg",
        },
        "type": "@shopgate/commerce-widgets/image",
        "width": 12,
      },
    ]
  }
>
  <div>
    <Image
      col={0}
      key="w0"
      row={0}
      settings={
        Object {
          "id": 83535,
          "image": "https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg",
        }
      }
      type="@shopgate/commerce-widgets/image"
      width={12}
    >
      <img
        alt=""
      />
    </Image>
  </div>
</Widgets>
`;

exports[`<Widgets /> should render a grid if the widget is of type grid 1`] = `
<Widgets
  components={
    Object {
      "@shopgate/commerce-widgets/image": [Function],
      "@shopgate/commerce-widgets/widget-grid": [Function],
    }
  }
  widgets={
    Array [
      Object {
        "settings": Object {
          "widgets": Array [
            Object {
              "col": 0,
              "height": 5,
              "row": 0,
              "settings": Object {
                "alt": "",
                "id": "84961",
                "image": "https://data.shopgate.com/shop_widget_images/23836/aedc545959f55e3f73851eca0ed40a75.min.jpeg",
                "link": "/category/",
              },
              "type": "@shopgate/commerce-widgets/image",
              "width": 12,
            },
          ],
        },
        "type": "@shopgate/commerce-widgets/widget-grid",
      },
    ]
  }
>
  <div>
    <WidgetGrid
      components={
        Object {
          "@shopgate/commerce-widgets/image": [Function],
          "@shopgate/commerce-widgets/widget-grid": [Function],
        }
      }
      config={
        Array [
          Object {
            "col": 0,
            "height": 5,
            "row": 0,
            "settings": Object {
              "alt": "",
              "id": "84961",
              "image": "https://data.shopgate.com/shop_widget_images/23836/aedc545959f55e3f73851eca0ed40a75.min.jpeg",
              "link": "/category/",
            },
            "type": "@shopgate/commerce-widgets/image",
            "width": 12,
          },
        ]
      }
      key="w0"
    >
      <div
        className={
          Object {
            "data-css-jldvt6": "",
          }
        }
      >
        <Widget
          component={[Function]}
          config={
            Object {
              "col": 0,
              "height": 5,
              "row": 0,
              "settings": Object {
                "alt": "",
                "id": "84961",
                "image": "https://data.shopgate.com/shop_widget_images/23836/aedc545959f55e3f73851eca0ed40a75.min.jpeg",
                "link": "/category/",
              },
              "type": "@shopgate/commerce-widgets/image",
              "width": 12,
            }
          }
          key="w0"
        >
          <GridItem
            className="css-14cdwhd"
            component="div"
            grow={0}
            shrink={1}
          >
            <div
              className="css-14cdwhd"
            >
              <div>
                <Suspense
                  fallback={<Loading />}
                >
                  <Image
                    ratio={
                      Array [
                        12,
                        5,
                      ]
                    }
                    settings={
                      Object {
                        "alt": "",
                        "id": "84961",
                        "image": "https://data.shopgate.com/shop_widget_images/23836/aedc545959f55e3f73851eca0ed40a75.min.jpeg",
                        "link": "/category/",
                      }
                    }
                  >
                    <img
                      alt=""
                    />
                  </Image>
                </Suspense>
              </div>
            </div>
          </GridItem>
        </Widget>
      </div>
    </WidgetGrid>
  </div>
</Widgets>
`;

exports[`<Widgets /> should render only one widget when the second one is not published and third one is invalid 1`] = `
<Widgets
  components={
    Object {
      "@shopgate/commerce-widgets/image": [Function],
      "@shopgate/commerce-widgets/widget-grid": [Function],
    }
  }
  widgets={
    Array [
      Object {
        "col": 0,
        "row": 0,
        "settings": Object {
          "id": 835351,
          "image": "https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg",
        },
        "type": "@shopgate/commerce-widgets/image",
        "width": 12,
      },
      Object {
        "col": 0,
        "row": 0,
        "settings": Object {
          "id": 835352,
          "image": "https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg",
        },
        "type": "@shopgate/commerce-widgets/imagefoo",
        "width": 12,
      },
      Object {
        "col": 0,
        "row": 0,
        "settings": Object {
          "id": 835353,
          "image": "https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg",
          "published": false,
        },
        "type": "@shopgate/commerce-widgets/image",
        "width": 12,
      },
    ]
  }
>
  <div>
    <Image
      col={0}
      key="w0"
      row={0}
      settings={
        Object {
          "id": 835351,
          "image": "https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg",
        }
      }
      type="@shopgate/commerce-widgets/image"
      width={12}
    >
      <img
        alt=""
      />
    </Image>
  </div>
</Widgets>
`;
